src/plugins/expressions/common/expression_functions/specs/tests/moving_average.test.ts

```diff
@@ -0,0 +1,474 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { functionWrapper } from './utils';
+import { movingAverage, MovingAverageArgs } from '../moving_average';
+import { ExecutionContext } from '../../../execution/types';
+import { Datatable } from '../../../expression_types/specs/datatable';
+
+describe('interpreter/functions#movingAverage', () => {
+  const fn = functionWrapper(movingAverage);
+  const runFn = (input: Datatable, args: MovingAverageArgs) =>
+    fn(input, args, {} as ExecutionContext) as Datatable;
+
+  it('calculates movingAverage', () => {
+    const result = runFn(
+      {
+        type: 'datatable',
+        columns: [{ id: 'val', name: 'val', meta: { type: 'number' } }],
+        rows: [{ val: 5 }, { val: 7 }, { val: 3 }, { val: 2 }],
+      },
+      { inputColumnId: 'val', outputColumnId: 'output' }
+    );
+    expect(result.columns).toContainEqual({
+      id: 'output',
+      name: 'output',
+      meta: { type: 'number' },
+    });
+    expect(result.rows.map((row) => row.output)).toEqual([
+      undefined,
+      5,
+      (5 + 7) / 2,
+      (5 + 7 + 3) / 3,
+    ]);
+  });
+
+  it('skips null or undefined values until there is real data', () => {
+    const result = runFn(
+      {
+        type: 'datatable',
+        columns: [{ id: 'val', name: 'val', meta: { type: 'number' } }],
+        rows: [
+          {},
+          { val: null },
+          { val: undefined },
+          { val: 1 },
+          { val: 2 },
+          { val: undefined },
+          { val: undefined },
+          { val: 4 },
+          { val: 8 },
+        ],
+      },
+      { inputColumnId: 'val', outputColumnId: 'output' }
+    );
+    expect(result.columns).toContainEqual({
+      id: 'output',
+      name: 'output',
+      meta: { type: 'number' },
+    });
+    expect(result.rows.map((row) => row.output)).toEqual([
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      1,
+      (1 + 2) / 2,
+      (1 + 2) / 2,
+      (1 + 2) / 2,
+      (1 + 2 + 4) / 3,
+    ]);
+  });
+
+  it('treats 0 as real data', () => {
+    const result = runFn(
+      {
+        type: 'datatable',
+        columns: [{ id: 'val', name: 'val', meta: { type: 'number' } }],
+        rows: [
+          {},
+          { val: null },
+          { val: undefined },
+          { val: 1 },
+          { val: 2 },
+          { val: 0 },
+          { val: undefined },
+          { val: 0 },
+          { val: undefined },
+          { val: 0 },
+          { val: 8 },
+          { val: 0 },
+        ],
+      },
+      { inputColumnId: 'val', outputColumnId: 'output', window: 3 }
+    );
+    expect(result.rows.map((row) => row.output)).toEqual([
```

Comment:
I tried to replicate this using Elasticsearch and it seems like it's doing something very different. This is what I tried (maybe I have made a mistake in here):
<details>
  <summary>Click to expand!</summary>
  
```
DELETE timeindex

POST timeindex/_doc
{
  "ts": "2020-09-30"
}

POST timeindex/_doc
{
  "ts": "2020-10-01"
}

POST timeindex/_doc
{
  "ts": "2020-10-02"
}

POST timeindex/_doc
{
  "ts": "2020-10-03",
  "val": 1
}

POST timeindex/_doc
{
  "ts": "2020-10-04",
  "val": 2
}

POST timeindex/_doc
{
  "ts": "2020-10-05",
  "val": 0
}


POST timeindex/_doc
{
  "ts": "2020-10-06"
}


POST timeindex/_doc
{
  "ts": "2020-10-07",
  "val": 0
}


POST timeindex/_doc
{
  "ts": "2020-10-08"
}


POST timeindex/_doc
{
  "ts": "2020-10-09",
  "val": 0
}


POST timeindex/_doc
{
  "ts": "2020-10-10",
  "val": 8
}


POST timeindex/_doc
{
  "ts": "2020-10-11",
  "val": 0
}


POST timeindex/_search
{
  "size": 0,
  "aggs": {
    "time": {
      "date_histogram": {
        "field": "ts",
        "interval": "day"
      },
      "aggs": {
        "sum": {
          "avg": {
              "field": "val"
          }
        },
        "deri": {
          "moving_fn": {
            "buckets_path": "sum",
            "window": 3,
            "script": "MovingFunctions.unweightedAvg(values)"
          }
        }
      }
    }
  }
}
```
</details>

Running this results in the following series:
| Values        | Moving avg           | Expression function |
| ---------- |:-------------:|-------------:|
| undefined      | undefined | undefined |
| undefined      | undefined | undefined |
| undefined      | undefined | undefined |
| 1      | undefined | undefined |
| 2      | 1.0 | 1 |
| 0      | 1.5 | 1.5 |
| undefined      | undefined | 1 |
| 0      | 1.0 | 1 |
| undefined      | undefined |0 |
| 0      | 0.66666 |0|
| 8      | 0 |0|
| 0      | 2.6666 |4|

It seems like the major difference is: 
* Elasticsearch: If an undefined value is encountered, skip the current row and do not change the window
* Expression function: If an undefined value is encountered, calculate the moving average based on the existing window and add "undefined" to window (which might push real values out)